File: 1239-maximum-length-of-a-concatenated-string-with-unique-characters/1239-maximum-length-of-a-concatenated-string-with-unique-characters.cpp
class Solution {
public:
    int used[26];
    bool isValid(string &s){
        unordered_set<char>st;
        for(char i : s)
            st.insert(i);
        return st.size() == s.size();
    }
    
    bool check(string s){
        for(char i : s){
            if(used[i - 'a'] + 1 > 1)   return false;
        }
        return true;
    }
    
    void mark(string s, bool val){
        for(char i : s){
            if(val)     ++used[i - 'a'];
            else    --used[i - 'a'];
        }
    }
    
    int helper(vector<string>& arr, int idx){
        if(idx >= arr.size())   return 0;
        
        int take = 0, skip = 0;
        
        if(isValid(arr[idx]) && check(arr[idx])){
            skip = helper(arr, idx + 1);
            mark(arr[idx], true);
            take = arr[idx].size() + helper(arr, idx + 1);
            mark(arr[idx], false);
        }else{
            skip = helper(arr, idx + 1);
        }
            
        return max(take, skip);
    }
    
    int maxLength(vector<string>& arr) {
        for(int i = 0; i < 26; i++)
            used[i] = 0;
        return helper(arr, 0);
    }
};

File: 1239-maximum-length-of-a-concatenated-string-with-unique-characters/1239-maximum-length-of-a-concatenated-string-with-unique-characters.java
class Solution {
    private int[] used = new int[26];

    private boolean isValid(String s) {
        Set<Character> set = new HashSet<>();
        for (char i : s.toCharArray()) {
            set.add(i);
        }
        return set.size() == s.length();
    }

    private boolean check(String s) {
        for (char i : s.toCharArray()) {
            if (used[i - 'a'] + 1 > 1) {
                return false;
            }
        }
        return true;
    }

    private void mark(String s, boolean val) {
        for (char i : s.toCharArray()) {
            if (val) {
                ++used[i - 'a'];
            } else {
                --used[i - 'a'];
            }
        }
    }

    private int helper(List<String> arr, int idx) {
        if (idx >= arr.size()) {
            return 0;
        }

        int take = 0, skip = 0;

        if (isValid(arr.get(idx)) && check(arr.get(idx))) {
            skip = helper(arr, idx + 1);
            mark(arr.get(idx), true);
            take = arr.get(idx).length() + helper(arr, idx + 1);
            mark(arr.get(idx), false);
        } else {
            skip = helper(arr, idx + 1);
        }

        return Math.max(take, skip);
    }

    public int maxLength(List<String> arr) {
        for (int i = 0; i < 26; i++) {
            used[i] = 0;
        }
        return helper(arr, 0);
    }
}